### PR TITLE
Upgrade Pester tests in Microsoft.PowerShell.Security folder to dashed v4 syntax

### PR DESCRIPTION
**Pull request by KevinMarquette**
## PR Summary
Fixes subset of #6245 
Use new dashed Pester v4 operators for tests in `Microsoft.PowerShell.Security`

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [NA] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [NA] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [NA] Issue filed - Issue link:
- **Testing - New and feature**
    - [X] Not Applicable or can only be tested interactively
    - **OR**
    - [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [NA] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 [hunk 1]
```diff
@@ -48,18 +48,18 @@ Describe "Certificate Provider tests" -Tags "CI" {
             param([string] $path)
             $expectedResolvedPath = Resolve-Path -LiteralPath $path
             $result = Get-Item -LiteralPath $path
-            $result | should not be null
```
Comment (KevinMarquette) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: The old test logic here is bad. Correcting the null test is causing the tests to fail.

Comment (KevinMarquette) [reply test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: The core issue here is that `Get-Item Cert:\` returns nothing but the test is written like it expects an object back. I am setting this test to pending for now.

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 [hunk 2]
```diff
@@ -44,22 +44,22 @@ Describe "Certificate Provider tests" -Tags "CI" {
     }
 
     Context "Get-Item tests" {
-        it "Should be able to get a certificate store, path: <path>" -TestCases $testLocations {
+        it "Should be able to get a certificate store, path: <path>" -TestCases $testLocations -Pending {
```
Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: Is the test failing because the test is malformed, or because there is an actual problem with code it is testing? If the test is failing due bad test logic, then pending is OK But if the target code for the test is is causing the tail failure, then the test should not be marked pending.

Comment (KevinMarquette) [reply test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: I'm going to correct the test logic so it only tests how the code currently works today. Then I will open an issue on the target code. I feel like it is an inconsistency in the target code, but low impact because it is an edge case and is unchanged from 5.1.

File: test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1 [hunk 3]
```diff
@@ -3,18 +3,18 @@
...
+        $ACL | Should -BeOfType "System.Security.AccessControl.DirectorySecurity"
     }
     It "Set-Acl can set the ACL of a directory" -pending {
         Setup -d testdir
         $directory = "$TESTDRIVE/testdir"
         $acl = get-acl $directory
         $accessRule = [System.Security.AccessControl.FileSystemAccessRule]::New("Everyone","FullControl","ContainerInherit,ObjectInherit","None","Allow")
         $acl.AddAccessRule($accessRule)
-        { $acl | Set-Acl $directory } | should not throw
+        $acl | Set-Acl $directory
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1]: I'd expect `{ $acl | Set-Acl $directory } | Should -Not -Throw`

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 (hunk 1 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: @KevinMarquette Please don't mess functional and style changes in future - this slows down the review very much. Open new issue and fix bugs in other PRs.

Here please revert line 47. I believe a bug is in line 50 - it should be `$result = Get-ChildItem -LiteralPath $path`.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 4]
```diff
@@ -8,23 +8,23 @@ Describe "ExecutionPolicy" -Tags "CI" {
...
         It "Should throw PlatformNotSupported when not on Windows" -Skip:$IsWindows {
-            { Set-ExecutionPolicy Unrestricted } | Should Throw "Operation is not supported on this platform."
+            { Set-ExecutionPolicy Unrestricted } | Should -Throw "Operation is not supported on this platform."
         }
 
         It "Should succeed on Windows" -Skip:($IsLinux -Or $IsMacOS) {
             # We use the Process scope to avoid affecting the system
             # Unrestricted is assumed "safe", otherwise these tests would not be running
-            { Set-ExecutionPolicy -Force -Scope Process -ExecutionPolicy Unrestricted } | Should Not Throw
+            Set-ExecutionPolicy -Force -Scope Process -ExecutionPolicy Unrestricted
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: Please fix:
`{ Set-ExecutionPolicy -Force -Scope Process -ExecutionPolicy Unrestricted } | Should -Not -Throw`

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 5]
```diff
@@ -688,7 +688,7 @@ ZoneId=$FileType
                 }
                 else
                 {
-                    $testScript | Should Not throw
+                    & $testScript
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: I believe it should be: `{ & $testScript } | Should -Not -Throw`

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 6]
```diff
@@ -773,7 +773,7 @@ ZoneId=$FileType
                     $scriptName=$testScript
 
                     $scriptResult = $null
-                    $exception = $null
+                    $exception    = $null
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: The change is unrelated  with the PR - please revert.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 7]
```diff
@@ -787,19 +787,19 @@ ZoneId=$FileType
                     $errorType = $null
                     if($null -ne $exception)
                     {
-                        $errorType = $exception.exception.getType()
+                        $errorType    = $exception.exception.getType()
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: The change is unrelated with the PR - please revert.
Please don't make formatting changes in future.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 8]
```diff
@@ -787,19 +787,19 @@ ZoneId=$FileType
...
                         $scriptResult = $null
                     }
                     $result = @{
-                        "result" = $scriptResult
+                        "result"    = $scriptResult
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: The change is unrelated with the PR - please revert.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 9]
```diff
@@ -787,19 +787,19 @@ ZoneId=$FileType
...
                         "exception" = $errorType
                     }
 
                     $actualResult = $result."result"
-                    $actualError = $result."exception"
+                    $actualError  = $result."exception"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: The change is unrelated with the PR - please revert.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 10]
```diff
@@ -787,19 +787,19 @@ ZoneId=$FileType
...
 
-                    $actualResult |  Should be $expected
-                    $actualError | Should be $error
+                    $actualResult | Should -Be $expected
+                    $actualError  | Should -Be $error
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: Please remove added space - formatting change.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 11]
```diff
@@ -933,7 +933,7 @@ ZoneId=$FileType
                 }
                 else
                 {
-                    $testScript | Should Not throw
+                    & $testScript
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: It should be `{ & $testScript } | Should -Not -Throw`

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 12]
```diff
@@ -1025,14 +1025,14 @@ ZoneId=$FileType
...
-                $testScript | should exist
+                $testScript | Should -Exist
                 if($error)
                 {
                     {& $testScript} | ShouldBeErrorId $error
                 }
                 else
                 {
-                    {& $testScript} | Should Not throw
+                    {& $testScript}
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: It should be `{ & $testScript } | Should -Not -Throw`

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 [hunk 13]
```diff
@@ -44,22 +44,22 @@ Describe "Certificate Provider tests" -Tags "CI" {
...
                 $resolvedPath = Resolve-Path $_.PSPath
-                $resolvedPath.Provider | should be $expectedResolvedPath.Provider
-                $resolvedPath.ProviderPath.TrimStart('\') | should be $expectedResolvedPath.ProviderPath.TrimStart('\')
+                $resolvedPath.Provider | Should -Be $expectedResolvedPath.Provider
+                $resolvedPath.ProviderPath.TrimStart('\') | Should -Be $expectedResolvedPath.ProviderPath.TrimStart('\')
             }
         }
         it "Should return two items at the root of the provider" {
-            (Get-Item -Path cert:\*).Count | should be 2
+            (Get-Item -Path cert:\*).Count | Should -Be 2
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: Please re-look the PR - we should use new syntax `Should -HaveCount`.

https://github.com/pester/Pester/blob/6cbac753e9e382aa3b1c482a3424c99602c01960/en-US/about_Should.help.txt

**Review by iSazonov — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1 (hunk 3 above)
Comment (KevinMarquette) [reply test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1]: I'll revert it, but I believe that Pester now discourages the use of `$scriptBlock | Should -Not -Throw` because all lines should not throw.

**Comment by KevinMarquette** [comment]
@iSazonov I reverted all functional changes and formatting changes with the exception of some trailing whitespace removal in one file.

I'll work on creating issues for the functional changes that I discovered that should be corrected.

File: test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1 [hunk 14]
```diff
@@ -11,7 +11,7 @@ Describe "Acl cmdlets are available and operate properly" -Tag CI {
         $acl = get-acl $directory
         $accessRule = [System.Security.AccessControl.FileSystemAccessRule]::New("Everyone","FullControl","ContainerInherit,ObjectInherit","None","Allow")
         $acl.AddAccessRule($accessRule)
-        $acl | Set-Acl $directory
+        { $acl | Set-Acl $directory } | Should Not -Throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1]: Typo Not -> -Not

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 15]
```diff
@@ -24,7 +24,7 @@ Describe "ExecutionPolicy" -Tags "CI" {
         It "Should succeed on Windows" -Skip:($IsLinux -Or $IsMacOS) {
             # We use the Process scope to avoid affecting the system
             # Unrestricted is assumed "safe", otherwise these tests would not be running
-            Set-ExecutionPolicy -Force -Scope Process -ExecutionPolicy Unrestricted
+            { Set-ExecutionPolicy -Force -Scope Process -ExecutionPolicy Unrestricted } | -Should -Not -Throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: Typo -Should -> Should.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 16]
```diff
@@ -688,7 +688,7 @@ ZoneId=$FileType
                 }
                 else
                 {
-                    & $testScript
+                    $testScript  | Should -Not -Throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: { & $testScript }  | Should -Not -Throw

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 [hunk 17]
```diff
@@ -933,7 +933,7 @@ ZoneId=$FileType
                 }
                 else
                 {
-                    & $testScript
+                    $testScript | Should Not throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: { & $testScript } | Should -Not -Throw

**Comment by iSazonov** [comment]
I added some comment but GitHub shows its as outdated.

**Comment by KevinMarquette** [comment]
Thank you. I'll make sure to review them.

**Comment by bergmeister** [comment]
I just wanted to say that I find it great that multiple people in the community help with doing this big upgrade!

**Comment by KevinMarquette** [comment]
Thank you for your patience.

Typos corrected and code changes other than the Pester parameters should be reverted. I created an issue for the functional changes.

**Comment by iSazonov** [comment]
@KevinMarquette Not all my comment addressed.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 (hunk 17 above)
Comment (KevinMarquette) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: this has been resolved

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 (hunk 16 above)
Comment (KevinMarquette) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: this has been resolved

File: test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1 (hunk 15 above)
Comment (KevinMarquette) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1]: this has been resolved

File: test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1 (hunk 14 above)
Comment (KevinMarquette) [reply test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1]: this has been resolved

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@kalgiz @adityapatwardhan Could you please review?

**Review by daxian-dbw — APPROVED** [review]
LGTM
